examples/stm32/f4/stm32f429i-discovery/ts-i2c/sdram.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * Copyright (C) 2014 Chuck McManis <cmcmanis@mcmanis.com>
+ *
+ * This library is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * This then is the initialization code extracted from the
+ * sdram example.
+ */
+#include <stdint.h>
+#include <libopencm3/stm32/gpio.h>
+#include <libopencm3/stm32/rcc.h>
+#include <libopencm3/stm32/fsmc.h>
+#include "clock.h"
+#include "sdram.h"
+
+/*
+ * This is just syntactic sugar but it helps, all of these
+ * GPIO pins get configured in exactly the same way.
+ */
+static struct {
+	uint32_t	gpio;
+	uint16_t	pins;
+} sdram_pins[6] = {
+	{GPIOB, GPIO5 | GPIO6 },
+	{GPIOC, GPIO0 },
+	{GPIOD, GPIO0 | GPIO1 | GPIO8 | GPIO9 | GPIO10 | GPIO14 | GPIO15},
+	{GPIOE, GPIO0 | GPIO1 | GPIO7 | GPIO8 | GPIO9 | GPIO10 |
+			GPIO11 | GPIO12 | GPIO13 | GPIO14 | GPIO15 },
+	{GPIOF, GPIO0 | GPIO1 | GPIO2 | GPIO3 | GPIO4 | GPIO5 | GPIO11 |
+			GPIO12 | GPIO13 | GPIO14 | GPIO15 },
+	{GPIOG, GPIO0 | GPIO1 | GPIO4 | GPIO5 | GPIO8 | GPIO15}
+};
+
+static struct sdram_timing timing = {
+	.trcd = 2,		/* RCD Delay */
+	.trp = 2,		/* RP Delay */
+	.twr = 2,		/* Write Recovery Time */
+	.trc = 7,		/* Row Cycle Delay */
+	.tras = 4,		/* Self Refresh Time */
+	.txsr = 7,		/* Exit Self Refresh Time */
+	.tmrd = 2,		/* Load to Active Delay */
+};
+
+/*
+ * Initialize the SD RAM controller.
+ */
+void
+sdram_init(void) {
+	int i;
+	uint32_t cr_tmp, tr_tmp; /* control, timing registers */
+
+	/*
+	* First all the GPIO pins that end up as SDRAM pins
+	*/
+	rcc_periph_clock_enable(RCC_GPIOB);
+	rcc_periph_clock_enable(RCC_GPIOC);
+	rcc_periph_clock_enable(RCC_GPIOD);
+	rcc_periph_clock_enable(RCC_GPIOE);
+	rcc_periph_clock_enable(RCC_GPIOF);
+	rcc_periph_clock_enable(RCC_GPIOG);
+
+	for (i = 0; i < 6; i++) {
+		gpio_mode_setup(sdram_pins[i].gpio, GPIO_MODE_AF,
+				GPIO_PUPD_NONE, sdram_pins[i].pins);
+		gpio_set_output_options(sdram_pins[i].gpio, GPIO_OTYPE_PP,
+					GPIO_OSPEED_50MHZ, sdram_pins[i].pins);
+		gpio_set_af(sdram_pins[i].gpio, GPIO_AF12, sdram_pins[i].pins);
+	}
+
+	/* Enable the SDRAM Controller */
+#if 1
+	rcc_periph_clock_enable(RCC_FSMC);
+#else
+	rcc_peripheral_enable_clock(&RCC_AHB3ENR, RCC_AHB3ENR_FMCEN);
+#endif
```

Comment:
no idea, i did not write sdram.c and have no idea why that is there. again, ad verbatim copy of lcd-serial.